## 4.3 오류 전달 방법
- 명시적인 방법
    - 코드를 호출한 쪽에서 오류가 발생할수 있음을 인지할 수밖에 없도록
    - 오류에 대한 처리는 호출하는 쪽에 위임
    - 오류를 모르고 넘어갈 수 있는 방법은 거의 없다.
- 암시적 방법
    - 오류를 알리지만 호출하는 쪽에서 그 오류를 신경쓰지 않아도 된다.
    - 문서나 코드를 읽는 등의 적극적인 노력이 필요
- 요란하게 실패할지 아니면 조용히 실패할지가 아니다.
    - 명시적인 기법을 통해 : 알아야 할 경우를 인지하게끔 하기
    - 암묵적인 기법을 통해 : 할 수 있는 것이 아무것도 없는 상황이라면 이런 상황을 처리해야 하는 부담에서 벗어나게 하는것
- 명시적 및 암묵적인 오류 전달 기법

### 4.3.1 요약: 예외
- 예외 : 코드에서 오류나 예외적인 상황이 발생한 경우 이를 전달하기 위한 방법
- 자바는 검사 예외(checked exception)와 비검사 예외(unchecked exception)개념이 있음

### 4.3.2 명시적 방법: 검사 예외(checked exception)
- 호출하는 쪽에서 예외를 인지하도록
- 예외 처리를 위한 코드를 작성하거나 자신의 함수 시그니처에 해당하는 예외 발생을 선언
- IOException, FileNotFoundException 등
- 검사 예외를 사용한 오류 전달
```java
class NegativeNumberException extends Exception { // 명시적 예외
    private final Double erroneousNumber;
    
    NegativeNumberException(Double erroneousNumber) { this.erroneousNumber = erroneousNumber;
    }
    Double getErroneousNumber() {
        return erroneousNumber;
    } 
}

Double getSquareRoot(Double value) throws NegativeNumberException { // 예외가 발생할 수 있음을 선언 필수
    if (value < 0.0) {
        throw new NegativeNumberException(value);
    }
    return Math.sqrt(value);
}
```

- 검사 예외 처리
```java
void displaySquareRoot() {
    Double value = ui.getInputNumber();
    try {
        ui.setOutput("Square root is: " + getSquareRoot(value));
    } catch (NegativeNumberException e) { // 예외 포착 (catch)
        ui.setError("Can't get square root of negative number: " + e.getErroneousNumber());
    }
}
```

- 예외를 시그니처에 선언하지 않거나 얘외처리를 안하면 컴파일 조차 안됨 > 명시적

### 4.3.3 암시적 방법: 비검사 예외(unchecked exception)
- 코드가 예외를 발생시킬 수 있다는 사실을 전혀 모를 수 있다.
- 문서화를 하더라도 코드 계약의 세부 조항이므로 오류를 암시적으로 알리는 방법
- NullPointerException, ClasssCastException 등
- 비검사 예외를 사용한 오류 전달
```java
class NegativeNumberException extends RuntimeException { // 암시적 예외
  private final Double erroneousNumber;

  NegativeNumberException(Double erroneousNumber) {
    this.erroneousNumber = erroneousNumber;
  }

  Double getErroneousNumber() {
    return erroneousNumber;
  }
}

/**
* @throws NegativeNumberException if the value is negative // 문서화
*/
Double getSquareRoot(Double value) {
    if (value < 0.0) { 
        throw new NegativeNumberException(value); // 비검사 예외 발생
    }
    return Math.sqrt(value);
}
```

- 비검사 예외 처리
```java
void displaySquareRoot() {
    Double value = ui.getInputNumber();
    try {
        ui.setOutput("Square root is: " + getSquareRoot(value));
    } catch (NegativeNumberException e) { // 예외 포착 (catch)
        ui.setError("Can't get square root of negative number: " + e.getErroneousNumber());
    } 
}
```

### 4.3.4 명시적 방법: 널값이 가능한 반환 유형
- 널 : 특정값을 계산하거나 얻는 것이 불가능함을 나타내기 위한 효과적이고 간단한 방법
- 언어가 널 안정성을 지원하는 경우 반환될 수 있다는 것을 알리고, 지원하지 않는 경우 옵셔널 반환 유형을 사용
```java
Double? getSquareRoot(Double value) {
    if(value< 0.0){
        return null; // 오류 발생시 널 리턴
    }
    return Math.sqrt(value);
}
```

```java
void displaySquareRoot() {
    Double? squareRoot = getSquareRoot(ui.getInputNumber());
    if (squareRoot == null) { // 강제적으로 확인해야 한다.
        ui.setError("Can't get square root of a negative number"); 
    } else {
        ui.setOutput("Square root is: " + squareRoot);
    } 
}
```

### 4.3.5 명시적 방법: 리절트 반환 유형
- 호출자에게 값을 얻을 수 없음을 알릴 뿐만 아니라 값을 얻을 수 없는 이유까지 알려주면 유용할때 사용
- 언어마다 지원하는 경우가 있고, 지원이 안될경우 번거롭지만 구현해야 함
```java
class Result<V, E> {
    private final Optional<V> value; // 성공시 값
    private final Optional<E> error; // 실패시 오류
    ...
}
```
```java
class NegativeNumberError extends Error {
  private final Double erroneousNumber;
  NegativeNumberError(Double erroneousNumber) {
    this.erroneousNumber = erroneousNumber;
  }
}

Result<Double, NegativeNumberError> getSquareRoot(Double value) { 
    if (value < 0.0) {
        return Result.ofError(new NegativeNumberError(value)); 
    }
  return Result.ofValue(Math.sqrt(value));
}
```

```java
void displaySquareRoot() {
    Result<Double, NegativeNumberError> squareRoot = getSquareRoot(ui.getInputNumber());
    if (squareRoot.hasError()) {
        ui.setError("Can't get square root of a negative number: " + squareRoot.getError().getErroneousNumber());
    } else {
        ui.setOutput("Square root is: " + squareRoot.getValue());
    }
}

```

### 4.3.6 명시적 방법: 아웃컴 반환 유형
- 어떤 일을 하는 동안 오류가 발생할 수 있고 그것을 호출한 쪽에 알리고자 한다면, 함수가 수행한 동작의 결과를 나타내는 값을 반환하도록 함수를 수정
```java
Boolean sendMessage(Channel channel, String message) {
    if (channel.isOpen()) {
        channel.send(message);
        return true; // 결과(성공) 전달
    }
    return false; // 결과(실패) 전달
}
```

```java
void sayHello(Channel channel) {
    if (sendMessage(channel, "hello")) {
        ui.setOutput("Hello sent"); // 성공한 경우
    } else {
        ui.setError("Unable to send hello"); // 실패한 경우
    }
}
```
- 아웃컴이 무시되지 않도록 : 자바의 [@CheckReturnValue](https://www.tabnine.com/code/java/classes/javax.annotation.CheckReturnValue)

```java
@CheckReturnValue
Boolean sendMessage(Channel channel, String message) {
    ... 
}
```

### 4.3.7 암시적 방법: 프로미스 또는 퓨처
- 비동기적으로 실행하는 코드를 작성할때 프로미스(promise)나 퓨처(future)를 반환
- 함수 내에서 오류가 발생하면 프로미스는 거부(reject)되고 그렇지 않으면 값을 리턴한다.
- 오류가 발생하고 프로미스가 거부될 수 있음을 알려면 프로미스를 생성하는 함수의 세부 조항이나 구현 세부 사항을 확인해야 하기 때문에 암묵적 오류 전달 기법이다.
```java
class NegativeNumberError extends Error {
  ...
}

Promise<Double> getSquareRoot(Double value) async { // 비동기 임을 알린다.
    await Timer.wait(Duration.ofSeconds(1));
    if (value < 0.0) {
        throw new NegativeNumberError(value);
    }
    return Math.sqrt(value);
}
```

```java
void displaySquareRoot() {
    getSquareRoot(ui.getInputNumber())
        .then(squareRoot ->
            ui.setOutput("Square root is: " + squareRoot)) // 성공하면
        .catch(error ->
            ui.setError("An error occurred: " + error.toString())); // 실패하면
}
```

### 4.3.8 암시적 방법: 매직값 반환
- 특정 값(-1, 0) 등을 반환하여 오류임을 전달
- 예상을 벗어나는 결과를 가져올 수 있고 버그로 이어질 수 있다.